tools/releaser/main.go

```diff
@@ -0,0 +1,221 @@
+
+This utility is intended to handle many of the steps to release a new version.
+
+`)
+		flag.PrintDefaults()
+	}
+
+	flag.Parse()
+
+	if tag == "" {
+		return fmt.Errorf("ERROR: tag is required")
+	}
+
+	if !tagRegexp.MatchString(tag) {
+		return tagErr
+	}
+
+	var goVersionArgs []string
+	if goVersion != "" {
+		versionParts := strings.Split(goVersion, ".")
+		if len(versionParts) < 2 {
+			flag.Usage()
+			return errors.New("please provide a valid Go version")
+		}
+		if minorVersion, err := strconv.Atoi(versionParts[1]); err != nil {
+			return fmt.Errorf("%q is not a valid Go version", goVersion)
+		} else if minorVersion > 0 {
+			versionParts[1] = strconv.Itoa(minorVersion - 1)
+		}
+		goVersionArgs = append(goVersionArgs, "-go", goVersion, "-compat", strings.Join(versionParts, "."))
+	}
+
+	workspacePath := path.Join(repoRoot, "WORKSPACE")
+
+	log("Running pre-release checks:")
+	if skipUpgrades {
+		log("- SKIPPING: go update commands")
+	} else {
+		log("- initial go update commands")
+		initialCommands := []struct {
+			cmd  string
+			args []string
+		}{
+			{cmd: "go", args: []string{"get", "-t", "-u", "./..."}},
+			{cmd: "tools/mod-tidy", args: goVersionArgs},
+		}
+		for _, c := range initialCommands {
+			cmd := exec.CommandContext(ctx, c.cmd, c.args...)
+			cmd.Dir = repoRoot
+			if out, err := cmd.CombinedOutput(); err != nil {
+				return fmt.Errorf(
+					"ERROR: running %s:%w\n%s",
+					strings.Join(append([]string{c.cmd}, c.args...), " "),
+					err,
+					out,
+				)
+			}
+		}
+	}
+
+	workspace, err := os.OpenFile(workspacePath, os.O_RDWR, 0o644)
+	if err != nil {
+		return err
+	}
+	defer workspace.Close()
+
+	log("- gazelle")
+	cmd := exec.CommandContext(ctx, "bazel", "run", "//:gazelle")
+	cmd.Dir = repoRoot
+	if out, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf("ERROR: gazelle failed: %w\n%s", err, out)
+	}
+
+	log("- checking if repository is clean")
+	cmd = exec.Command("git", "diff", "--stat", "--exit-code")
+	cmd.Dir = repoRoot
+	if out, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf(
+			"ERROR: repository is dirty. Please commit your changes to continue:\n%s",
+			out,
+		)
+	}
+
+	log("Creating tag " + tag)
+
+	cmd = exec.Command("git", "tag", tag)
+	cmd.Dir = repoRoot
+	if out, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf("ERROR: failed to create git tag: %w\n%s", err, string(out))
+	}
+
+	log("Creating archive bazel-zig-cc-%s.tar", tag)
+
+	cmd = exec.Command(
+		"git",
+		append([]string{
+			"archive",
+			"--format=tar",
+			tag,
+		}, _paths...)...,
+	)
+	cmd.Dir = repoRoot
+
+	out, err := cmd.Output()
+	if err != nil {
+		var exitError *exec.ExitError
+		errors.As(err, &exitError)
+		return fmt.Errorf("ERROR: failed to create git archive: %w\n%s", err, exitError.Stderr)
+	}
+
+	log("Compressing bazel-zig-cc-%s.tar", tag)
+
+	tgz := zopfli.Gzip(out)
```

Comment:
If we use `compress/gzip`, then
1. we can stream the `StdoutPipe` directly into `gzip.Writer`
2. we have one less third-party dep to worry about and the potential duplicate definitionif zopfli repo

Reply:
Yes, it would be more convenient to use `compress/gzip`, but zopfli offers *better compression*.

It doesn't really matter right now (22kb vs 25kb). But I have set a precedent to myself years ago -- we should use the compression algorithm that makes sense. Assuming gzip:
- compress often, extract once -- `encoding/gzip` or more cpu-efficient alternatives (there are better zlib implementations)
- compress once, extract often -- zopfli.

I am slightly more inclined to accept a small maintenance burden to use better compression, but can totally be convinced that maintainability is more important.